src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+            break;
+        }
+        prev_id          = keyset.keyset_id;
+        keyset.keyset_id = keyset.next;
+        keyset_count++;
+    }
+
+    VerifyOrReturnError(keyset.keyset_id == target_id, CHIP_ERROR_KEY_NOT_FOUND);
+
+    keyset.Delete(mStorage);
+    if (keyset_count > 0)
+    {
+        // Remove intermediate keyset, update previous
+        KeysetData prev_data(fabric_index, prev_id);
+        ReturnErrorOnFailure(prev_data.Load(mStorage));
+        prev_data.next = keyset.next;
+        ReturnErrorOnFailure(prev_data.Save(mStorage));
+    }
+    else
+    {
+        // Remove first keyset
+        fabric.first_keyset = keyset.next;
+    }
+    if (fabric.keyset_count > 0)
+    {
+        fabric.keyset_count--;
+    }
+    // Update fabric info
+    return fabric.Save(mStorage);
+}
+
+GroupDataProvider::KeysetIterator * GroupDataProviderImpl::IterateKeysets(chip::FabricIndex fabric_index)
+{
+    VerifyOrReturnError(mInitialized, nullptr);
+    return mKeysetIterators.CreateObject(*this, fabric_index);
+}
+
+GroupDataProviderImpl::KeysetIteratorImpl::KeysetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index) :
+    mProvider(provider), mFabric(fabric_index)
+{
+    FabricData fabric(fabric_index);
+    if (CHIP_NO_ERROR == fabric.Load(provider.mStorage))
+    {
+        mNextId = fabric.first_keyset;
+        mCount  = fabric.keyset_count;
+        mIndex  = 0;
+    }
+}
+
+size_t GroupDataProviderImpl::KeysetIteratorImpl::Count()
+{
+    return mCount;
+}
+
+bool GroupDataProviderImpl::KeysetIteratorImpl::Next(Keyset & item)
+{
+    VerifyOrReturnError(mIndex < mCount, false);
+
+    KeysetData keyset(mFabric, mNextId);
+    VerifyOrReturnError(CHIP_NO_ERROR == keyset.Load(mProvider.mStorage), false);
+
+    mIndex++;
+    mNextId            = keyset.next;
+    item.keyset_id     = keyset.keyset_id;
+    item.policy        = keyset.policy;
+    item.num_keys_used = keyset.num_keys_used;
+    memcpy(item.epoch_keys, keyset.epoch_keys, sizeof(item.epoch_keys));
+    return true;
+}
+
+void GroupDataProviderImpl::KeysetIteratorImpl::Release()
+{
+    mProvider.mKeysetIterators.ReleaseObject(this);
+}
+
+//
+// Fabrics
+//
+
+CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
+{
+    FabricData fabric(fabric_index);
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_KEY_NOT_FOUND);
+
+    // Remove Group Mappings
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    do
+    {
+        if (CHIP_NO_ERROR != endpoint.Load(mStorage))
+        {
+            break;
+        }
+        RemoveAllGroupMappings(fabric_index, endpoint.id);
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+
+    // Remove States
+
+    StateListData states;
+
+    // Load state list info
+    states.Load(mStorage);
```

Comment:
Add a comment for why it's OK not to check error (e.g. `states.count` will stay 0 on error).

There is an unchecked assumption here: if `count` is not the last field loaded in `StateListData.Load` and there are prior errors, we may have states.count != 0, but error loading states.

The `StateListData.Load` must have a comment and code to ensure that if we change format, `count` is always still read as zero on error if you ignore return value... This is a bit of a complex invariant to keep "track of" over maintenance and time.

Reply:
I have added the comment. As for ensuring the consistency of the data, PersistentData has a pure virtual Clear() method that is called before de-serializing, and is used for this purpose. Nevertheless, I have moved the call to Clear() up in the Load() method, so it is also called before SyncGetKeyValue.